<h1 style='font-family:sans-serif;'>⚕️Apa itu <a href="https://www.python.org/" style='color:blue;'>Python?</a></h1>
<img src="https://media.tenor.com/hW_mTYy_zS4AAAAi/gojo-satoru.gif" height=250 width=250 alt="License">

<div style="background-color:#f5f5f5; border-left:5px solid #555; padding:12px;">
Python itu bahasa pemrograman yang simple, gampang dibaca, dan populer.
dipake buat banyak hal: bikin website, analisis data, machine learning, bikin game, sampai otomasi kerjaan sehari-hari.
</div>

<div style="background-color:#f5f5f5; border-left:5px solid #555; padding:12px;">
Kelebihannya python: <br>
1. sintaksnya gampang → mendekati bahasa inggris. <br>
2. librarynya banyak → tinggal pake, gak usah bikin dari nol. <br>
3. komunitas besar → gampang cari solusi kalo error. <br>
</div>

## ⚜️Python Syntax

<h4 style="color:gray">1. Indentasi.</h4>

<div style="background-color:#f5f5f5; border-left:5px solid #555; padding:12px;">
Python tidak menggunakan tanda kurung kurawal {} untuk menandai blok kode, melainkan indentasi (spasi atau tab).
Biasanya standar yang dipakai adalah 4 spasi. Kalau salah indentasi, Python akan error.
</div>

In [8]:
x = 5
if x > 3:
    print("x lebih besar dari 3")
else:
    print("x kecil atau sama dengan 3")

x lebih besar dari 3


<h4 style="color:gray">2. Case Sensitive</h4>

<div style="background-color:#f5f5f5; border-left:5px solid #555; padding:12px;">
Python membedakan huruf besar dan huruf kecil (case-sensitive). Artinya, nama, Nama, dan NAMA dianggap tiga variabel berbeda.
dibawah ini merupakan contoh yang salah karena konteks variabel nama dan Nama itu berbeda.
</div>

In [12]:
nama = "Budi"
print(Nama)

NameError: name 'Nama' is not defined

<h4 style="color:gray">3. Komentar</h4>

<div style="background-color:#f5f5f5; border-left:5px solid #555; padding:12px;">
Komentar digunakan untuk menulis catatan pada kode. Tidak akan dieksekusi.
</div>

In [15]:
# ini komentar satu baris
print("Halo Dunia")
#print("halo smeesta")

Halo Dunia


<div style="
    background-color:#ffffff;
    border-bottom:8px solid green;
    box-shadow: 0 2px 6px rgba(0,0,0,0.15);
    border-radius:10px;
    padding:16px;
    margin:12px 0;
    font-family: 'Segoe UI', sans-serif;
">

Syntax & Print Perintah Python yang benar untuk menampilkan teks "Halo UTM" adalah?
    
    a. echo("Halo UTM")
    b. printf("Halo UTM")
    c. cout << "Halo UTM"
    d. print("Halo UTM")

<div style="margin-top:10px; background-color:#f9fafb; border-radius:6px; padding:8px 12px; border:1px dashed #3b82f6;">
✅ <i>Jawaban: D</i>
</div> 

</div>

## ⚜️Variable

<div style="background-color:#f5f5f5; border-left:5px solid #555; padding:12px;">
Variabel itu sperti tempat nyimpen data. beda sperti bahasa pemrograman lain pada umumnya, python gak perlu deklarasi tipe data tapi langsung aja kasih valuenya.
</div>

In [3]:
nama = "Budi"
umur = 18
tinggi = 170.5

print(nama, umur, tinggi)

Budi 18 170.5


<div style="background-color:#f5f5f5; border-left:5px solid #555; padding:12px;">

<p style="font-family:tahoma;">🚀<b>Ketentuan Membuat Variabel Dalam Python:</b></p>

- Nama variabel harus diawali dengan huruf atau garis bawah (_), tidak boleh diawali dengan angka.  
- Variabel hanya boleh terdiri dari huruf, angka, dan garis bawah, tanpa spasi.  
- Nama variabel bersifat case-sensitive, artinya `data`, `Data`, dan `DATA` dianggap berbeda.  
- Penulisan nama variabel yang baik mengikuti aturan snake_case, misalnya `total_harga` atau `jumlah_siswa`.  
- Tidak boleh menggunakan kata kunci (keywords) Python seperti `if`, `while`, atau `class` sebagai nama variabel.  
- Sebaiknya nama variabel dibuat deskriptif agar mudah dipahami, misalnya `nilai_rata` lebih baik daripada `nr`.  

</div>

<div style="
    background-color:#ffffff;
    border-bottom:8px solid green;
    box-shadow: 0 2px 6px rgba(0,0,0,0.15);
    border-radius:10px;
    padding:16px;
    margin:12px 0;
    font-family: 'Segoe UI', sans-serif;
">
Mana variabel yang benar dalam Python?

    a. 8nama = "Vian"
    b. nama! = "Vian"
    c. nama = "Dafa"
    d. string nama = "Dafa"

<div style="margin-top:10px; background-color:#f9fafb; border-radius:6px; padding:8px 12px; border:1px dashed #3b82f6;">
✅ <i>Jawaban: D</i>
</div>

</div>

## ⚜️Error

<div style="background-color:#f5f5f5; border-left:5px solid #555; padding:12px;">
<h5>1. SyntaxError</h5>
    
Ini error yang paling sering ditemuin waktu baru belajar Python. Intinya, Python nggak ngerti kode kita karena ada yang salah nulis. Misalnya lupa nutup tanda kurung, salah pakai titik dua, atau struktur kodenya nggak sesuai aturan.
Contoh:
</div>

In [5]:
print("Error"

_IncompleteInputError: incomplete input (3904745168.py, line 1)

<div style="background-color:#f5f5f5; border-left:5px solid #555; padding:12px;">
<h5>2. IndentationError</h5>
    
Kalau di Python, indentasi (spasi/tab di awal baris) itu wajib hukumnya. Beda sama bahasa lain yang pakai { }. Jadi kalau indentasinya berantakan, pasti error.
Contoh:
</div>

In [4]:
if True:
print("Ini error")

IndentationError: expected an indented block after 'if' statement on line 1 (3598588081.py, line 2)

<div style="background-color:#f5f5f5; border-left:5px solid #555; padding:12px;">
<h5>3. NameError</h5>
    
Ini muncul kalau kita manggil variabel atau fungsi yang sebenernya nggak ada. Bisa karena lupa bikin variabelnya dulu, atau salah ketik.
Contoh:
</div>

In [6]:
print(total)

NameError: name 'total' is not defined

<div style="background-color:#f5f5f5; border-left:5px solid #555; padding:12px;">
<h5>4. TypeError</h5>
    
Error ini muncul kalau kita maksa ngelakuin operasi pada dua tipe data yang nggak cocok. Misalnya, mau nambahin angka sama string, jelas Python nggak ngerti.
Contoh:
</div>

In [7]:
angka = 5
print(angka + "10")

TypeError: unsupported operand type(s) for +: 'int' and 'str'

## ⚜️Tipe Data

<div style="background-color:#f5f5f5; border-left:5px solid #555; padding:12px;">
Python punya beberapa tipe data dasar yang sering dipakai untuk menyimpan informasi. Dengan mengenal tipe data, kita bisa tahu cara kerja sebuah variabel dan operasi apa saja yang bisa dilakukan.
</div>

<div style="background-color:#f5f5f5; border-left:5px solid #555; padding:12px;">
<h5>1. String</h5>
    
Digunakan untuk menyimpan teks. String ditulis dengan tanda kutip tunggal (‘ ’) atau ganda (“ ”).
</div>

In [1]:
teks = "Python UTM"
print(teks.upper())
print(teks[0])

PYTHON UTM
P


<div style="background-color:#f5f5f5; border-left:5px solid #555; padding:12px;">
<h5>2. Integer</h5>
    
Berisi angka bulat, bisa positif atau negatif.
</div>

In [2]:
angka = 25
print(angka + 5)

30



<div style="background-color:#f5f5f5; border-left:5px solid #555; padding:12px;">
<h5>3. Float</h5>
    
Angka pecahan atau desimal.
</div>

In [4]:
pi = 3.14
print(pi * 2)

6.28



<div style="background-color:#f5f5f5; border-left:5px solid #555; padding:12px;">
<h5>4. Boolean</h5>
    
Nilai logika yang hanya punya dua kemungkinan: True atau False.
</div>

In [5]:
status = True
print(10 > 5)

True


<div style="background-color:#f5f5f5; border-left:5px solid #555; padding:12px;">
<h5>5. List</h5>
    
Struktur data untuk menyimpan banyak nilai dalam satu variabel. List bisa diubah (mutable).
</div>

In [7]:
buah = ["apel", "mangga", "jeruk"]
buah.append("pisang")
print(buah[1])

mangga


<div style="
    background-color:#ffffff;
    border-bottom:8px solid green;
    box-shadow: 0 2px 6px rgba(0,0,0,0.15);
    border-radius:10px;
    padding:16px;
    margin:12px 0;
    font-family: 'Segoe UI', sans-serif;
">
Apa output dari kode berikut:


    angka = 7
    pecahan = 2.5
    print(angka + pecahan)

<div style="margin-top:10px; background-color:#f9fafb; border-radius:6px; padding:8px 12px; border:1px dashed #3b82f6;">
✅ <i>Jawabanan Dibawah</i>
</div>


In [8]:
angka = 7
pecahan = 2.5
print(angka + pecahan)

9.5


## ⚜️Operator

<div style="
    background-color:#ffffff;
    border-left:6px solid #0ea5e9;
    box-shadow: 0 2px 6px rgba(0,0,0,0.12);
    border-radius:10px;
    padding:16px;
    margin:8px 0;
    font-family: 'Segoe UI', sans-serif;
">

<b>📘 Pembahasan – Operator dalam Python</b>  

Operator adalah simbol yang digunakan untuk melakukan operasi pada data atau variabel.  
Secara umum, ada tiga jenis utama: **Aritmatika, Perbandingan, dan Logika**.  

---

🔹 <b>Aritmatika</b>  
- <code>+</code> → tambah  
- <code>-</code> → kurang  
- <code>*</code> → kali  
- <code>/</code> → bagi  
- <code>%</code> → sisa bagi (modulus)  
- <code>**</code> → pangkat  
- <code>//</code> → pembagian bulat  

---

🔹 <b>Perbandingan</b> (hasilnya <code>True</code> / <code>False</code>)  
- <code>==</code> → sama dengan  
- <code>!=</code> → tidak sama dengan  
- <code>></code> → lebih besar  
- <code><</code> → lebih kecil  
- <code>>=</code> → lebih besar atau sama dengan  
- <code><=</code> → lebih kecil atau sama dengan  

---

🔹 <b>Logika</b>  
- <code>and</code> → harus dua-duanya <code>True</code>  
- <code>or</code> → cukup salah satu <code>True</code>  
- <code>not</code> → kebalikannya  

</div>


In [9]:
print("-------------------")
#ARITMATIKA:

a = 10
b = 3

print(a + b)
print(a - b)   
print(a * b)   
print(a / b)   
print(a % b)   
print(a ** b)  
print(a // b)
print("-------------------")

#PERBANDINGAN:

x = 5
y = 8

print(x == y) 
print(x != y) 
print(x > y)
print("-------------------")

#LOGIKA:

a = True
b = False

print(a and b)
print(a or b)
print(not a)
print("-------------------")

-------------------
13
7
30
3.3333333333333335
1
1000
3
-------------------
False
True
False
-------------------
False
True
False
-------------------


In [10]:
x = 10
y = 3
z = 5

a = x + y * z
b = (x - z) // y
c = x % y
d = (x > y) and (z < x)
e = (x == 10) or (y == 5)
f = not (z == 5)
g = (x ** 2) / (y + z)
h = (x >= z) and not (y == 3)

<div style="border-left:4px solid #8b5cf6; padding:12px; margin:10px 0;">
<b>⚜️ Quiz Operator:</b><br>
Diketahui: <br>
<code>x = 10</code>, <code>y = 3</code>, <code>z = 5</code>

1. Tentukan hasil dari operasi berikut:<br>

a. <code>a = x + y * z</code><br>
b. <code>b = (x - z) // y</code><br>
c. <code>c = x % y</code><br>
d. <code>d = (x > y) and (z < x)</code><br>
e. <code>e = (x == 10) or (y == 5)</code><br>
f. <code>f = not (z == 5)</code><br>
g. <code>g = (x ** 2) / (y + z)</code><br>
h. <code>h = (x >= z) and not (y == 3)</code><br><br>

2. Sebutkan jenis operator apa saja yang dipakai (aritmatika, pembanding, logika).<br>
3. Jelaskan perbedaan antara <code>/</code>, <code>//</code>, dan <code>%</code>.<br>
</div>

<div style="border-left:4px solid #10b981; padding:12px; margin:10px 0;">
<b>✅ Jawaban Nomer 1 (Perhitungan):</b><br>
- a = x + y * z = 10 + 3 * 5 = 10 + 15 = <b>25</b><br>
- b = (x - z) // y = (10 - 5) // 3 = 5 // 3 = <b>1</b><br>
- c = x % y = 10 % 3 = <b>1</b><br>
- d = (x > y) and (z < x) = (10 > 3) and (5 < 10) = True and True = <b>True</b><br>
- e = (x == 10) or (y == 5) = (True) or (False) = <b>True</b><br>
- f = not (z == 5) = not (True) = <b>False</b><br>
- g = (x ** 2) / (y + z) = (100) / (8) = <b>12.5</b><br>
- h = (x >= z) and not (y == 3) = (10 >= 5) and not (True) = True and False = <b>False</b><br>
</div>

<div style="border-left:4px solid #f59e0b; padding:12px; margin:10px 0;">
<b>✅ Jawaban Nomer 2 (Jenis Operator):</b><br>
- <b>Aritmatika</b> → <code>+</code>, <code>-</code>, <code>*</code>, <code>//</code>, <code>%</code>, <code>**</code>, <code>/</code><br>
- <b>Perbandingan</b> → <code>></code>, <code><</code>, <code>==</code>, <code>>=</code><br>
- <b>Logika</b> → <code>and</code>, <code>or</code>, <code>not</code><br>
</div>

<div style="border-left:4px solid #ef4444; padding:12px; margin:10px 0;">
<b>✅ Jawaban Nomer 3 (Perbedaan <code>/</code>, <code>//</code>, <code>%</code>):</b><br>
- <b>/</b> → pembagian biasa, hasil bisa desimal. Contoh: <code>10 / 3 = 3.333...</code><br>
- <b>//</b> → pembagian bulat (*floor division*), hasil dibulatkan ke bawah. Contoh: <code>10 // 3 = 3</code><br>
- <b>%</b> → sisa pembagian (modulus). Contoh: <code>10 % 3 = 1</code><br>
</div>


## ⚜️If/Else

<div style="background-color:#f5f5f5; border-left:5px solid #555; padding:12px;">
Pada dasarnya, pernyataan if-else 🐍 adalah salah satu konsep fundamental dalam pemrograman yang berfungsi untuk mengendalikan alur program. Bayangkan seperti sebuah persimpangan di jalan. Anda bisa memilih jalan mana yang akan diambil berdasarkan kondisi tertentu. if-else memungkinkan program Anda "memilih" jalan yang berbeda berdasarkan hasil dari suatu kondisi logis.

Bagaimana if-else Bekerja?
Setiap kali Anda menggunakan if, Python akan mengevaluasi ekspresi kondisional (seperti umur >= 18). Ekspresi ini akan selalu menghasilkan salah satu dari dua nilai: True (Benar) atau False (Salah).

Jika ekspresi menghasilkan True, blok kode di bawah if akan dieksekusi.

Jika ekspresi menghasilkan False, Python akan melompati blok kode if dan beralih ke blok else (jika ada).

Penting untuk diingat bahwa indentasi (spasi di awal baris) sangat krusial dalam Python. Blok kode yang termasuk dalam if atau else harus memiliki level indentasi yang sama, biasanya 4 spasi.
</div>

In [1]:
suhu_air = 95

if suhu_air > 100:
    print("Airnya sudah mendidih!")
    print("Hati-hati, uapnya panas!")
else:
    print("Air belum mendidih.")

Air belum mendidih.


<div style="background-color:#f5f5f5; border-left:5px solid #555; padding:12px;">
Pada contoh di atas, karena suhu_air adalah 95, yang mana tidak lebih besar dari 100, kondisi if menjadi False. Oleh karena itu, Python akan mengeksekusi blok kode di dalam else
</div>

In [19]:
total = 18000

if total > 20000:
    print("Kamu mendapatkan diskon 20%")
elif total >= 13000:
    print("Kamu mendapatkan diskon 10%")
else:
    print("Kamu tidak mendapatkan diskon")


Kamu mendapatkan diskon 10%


<div style="background-color:#f5f5f5; border-left:5px solid #555; padding:12px;">
Seringkali, satu kondisi if-else tidak cukup. Di sinilah elif (singkatan dari "else if") berperan. elif memungkinkan Anda untuk memeriksa kondisi tambahan secara berurutan. Python akan mengecek setiap elif satu per satu jika kondisi sebelumnya salah.
</div>

In [2]:
nilai = 85

if nilai >= 90:
    print("Nilai Anda A, luar biasa!")
elif nilai >= 80:  
    print("Nilai Anda B, bagus!")
elif nilai >= 70:
    print("Nilai Anda C, lumayan.")
else:
    print("Nilai Anda D, perlu belajar lebih giat.")

Nilai Anda B, bagus!


<div style="background-color:#f5f5f5; border-left:5px solid #555; padding:12px;">
Dalam skenario ini, Python akan mengecek:

if nilai >= 90? False (karena 85 tidak lebih dari 90).

elif nilai >= 80? True (karena 85 lebih dari 80).

Karena kondisi elif ini True, Python akan menjalankan print("Nilai Anda B, bagus!") dan langsung keluar dari seluruh struktur if-elif-else ini. Kode di bawahnya (elif dan else yang lain) tidak akan pernah dieksekusi. Ini membuat alur program menjadi efisien.
</div>

### Kondisi Majemuk dan if Bersarang:

<div style="background-color:#f5f5f5; border-left:5px solid #555; padding:12px;">
Anda juga bisa membuat kondisi yang lebih kompleks menggunakan operator logis seperti and, or, dan not.

and: Kondisi menjadi True jika kedua sisi and adalah True.

or: Kondisi menjadi True jika salah satu sisi or adalah True.

not: Membalik nilai dari True menjadi False atau sebaliknya.
</div>

In [20]:
nilai = 90

if nilai >= 70:
    print("Lulus")
    if nilai == 100:
        print("Nilai sempurna!")
else:
    print("Tidak lulus")


Lulus


In [3]:
tinggi = 175
berat = 70

if tinggi > 170 and berat < 80:
    print("Anda memenuhi syarat untuk tim basket.")
else:
    print("Maaf, Anda tidak memenuhi syarat.")

Anda memenuhi syarat untuk tim basket.


<div style="background-color:#f5f5f5; border-left:5px solid #555; padding:12px;">
Selain itu, Anda bisa menggunakan if bersarang (nested if), yaitu menempatkan satu pernyataan if-else di dalam blok if atau else lainnya. Ini berguna untuk memeriksa kondisi yang lebih spesifik setelah kondisi utama terpenuhi.
</div>

In [5]:
jenis_kelamin = "Laki-laki"
usia = 25

if jenis_kelamin == "Laki-laki":
    if usia >= 18:
        print("Selamat datang, Tuan!")
    else:
        print("Anda adalah laki-laki di bawah umur.")
else:
    print("Anda bukan laki-laki.")

Selamat datang, Tuan!


<div style="background-color:#f5f5f5; border-left:5px solid #555; padding:12px;">
Penggunaan if-else adalah fondasi untuk menciptakan program yang dinamis dan cerdas, yang dapat merespons berbagai input atau kondisi.
</div>

### Quiz

<div style="
    background-color:#ffffff;
    border-bottom:8px solid green;
    box-shadow: 0 2px 6px rgba(0,0,0,0.15);
    border-radius:10px;
    padding:16px;
    margin:12px 0;
    font-family: 'Segoe UI', sans-serif;
">
<b>Apa output dari kode berikut:</b>
<br><br>

    angka = 25
    if angka % 2 == 0:
        print("Angka genap")
    else:
        print("Angka ganjil")

<div style="margin-top:10px; background-color:#f9fafb; border-radius:6px; padding:8px 12px; border:1px dashed #3b82f6;">
✅ <i>Angka Ganjil</i>
</div>


<div style="
    background-color:#ffffff;
    border-bottom:8px solid green;
    box-shadow: 0 2px 6px rgba(0,0,0,0.15);
    border-radius:10px;
    padding:16px;
    margin:12px 0;
    font-family: 'Segoe UI', sans-serif;
">
<b>Dalam struktur percabangan `if-else` di Python, kapan blok kode di dalam `else` akan dieksekusi?</b>
<br><br>

    A. Ketika kondisi pada pernyataan `if` bernilai `False`.
    B. Ketika kondisi pada pernyataan `if` bernilai `True`.
    C. Setiap saat, terlepas dari nilai kondisi `if`.
    D. Ketika terjadi kesalahan (error) pada kode `if`.

<div style="margin-top:10px; background-color:#f9fafb; border-radius:6px; padding:8px 12px; border:1px dashed #3b82f6;">
✅ <i>A. Ketika kondisi pada pernyataan 'if' bernilai False</i>
</div>


## ⚜️Pengulangan

<div style="background-color:#f5f5f5; border-left:5px solid #555; padding:12px;">
    <b>Apa Itu Perulangan (Looping)?</b> Secara sederhana, perulangan adalah instruksi yang meminta komputer untuk mengulang satu blok kode berulang kali. Bayangkan Anda harus menyalin 100 baris data dari satu file ke file lain. Menuliskan perintah salin sebanyak 100 kali adalah hal yang tidak praktis. Perulangan memungkinkan Anda untuk menuliskan perintah salin satu kali dan memberitahu program untuk mengulanginya 100 kali.

Di Python, ada dua jenis perulangan utama yang menjadi fondasi dari semua perulangan lainnya.
</div>

<div style="background-color:#f5f5f5; border-left:5px solid #555; padding:12px;">
<b>1. Perulangan for</b> <br>
Perulangan for digunakan untuk iterasi yang pasti (definite iteration). Ini artinya, Anda tahu persis berapa kali perulangan akan berjalan, biasanya dengan mengulang setiap elemen dalam sebuah urutan (sequence). Urutan ini bisa berupa daftar, tuple, string, atau objek lain yang dapat diiterasi.
<br>
Iterasi di dalam Urutan
for sangat kuat karena ia akan secara otomatis mengambil setiap elemen dari urutan dan memberikannya ke variabel yang Anda tentukan.
</div>

In [24]:
buah = ["apel", "mangga", "jeruk"]

for b in buah:
    print("Aku suka", b)

print('-------------') #atau pake range

for i in range(5):
    print("Perulangan ke-", i)

Aku suka apel
Aku suka mangga
Aku suka jeruk
-------------
Perulangan ke- 0
Perulangan ke- 1
Perulangan ke- 2
Perulangan ke- 3
Perulangan ke- 4


In [6]:
daftar_belanja = ["apel", "roti", "susu"]
for item in daftar_belanja:
    print(f"Saya akan beli {item}.")

Saya akan beli apel.
Saya akan beli roti.
Saya akan beli susu.


<div style="background-color:#f5f5f5; border-left:5px solid #555; padding:12px;">
<b>Menggunakan range() untuk Iterasi Berdasarkan Angka</b>
<br>
Fungsi range() adalah teman baik perulangan for saat Anda ingin mengulang berdasarkan angka. Ia membuat urutan angka yang tidak terlihat (lazily generated).

range(stop): Menghasilkan angka dari 0 hingga stop-1.

range(start, stop): Menghasilkan angka dari start hingga stop-1.

range(start, stop, step): Menghasilkan angka dari start hingga stop-1, dengan lompatan sebesar step.
</div>

In [7]:
for i in range(3):
    print("Selamat datang!")

for angka in range(2, 10, 2):
    print(f"Angka genap: {angka}")

Selamat datang!
Selamat datang!
Selamat datang!
Angka genap: 2
Angka genap: 4
Angka genap: 6
Angka genap: 8


<div style="background-color:#f5f5f5; border-left:5px solid #555; padding:12px;">
<b>2. While</b>, jalan terus selama kondisi masih True. Perulangan while digunakan untuk iterasi yang tidak pasti (indefinite iteration). Perulangan ini akan terus berjalan selama sebuah kondisi bernilai True. Perulangan ini sangat berguna ketika Anda tidak tahu kapan perulangan akan berakhir.

Mengandalkan Kondisi Logis
Hal yang paling penting dari perulangan while adalah memastikan ada sesuatu di dalam blok kode yang akan membuat kondisi menjadi False di kemudian hari.

<br>

<b>Mengandalkan Kondisi Logis</b><br>
Hal yang paling penting dari perulangan while adalah memastikan ada sesuatu di dalam blok kode yang akan membuat kondisi menjadi False di kemudian hari.
</div>

In [8]:
jumlah_tiket = 5
while jumlah_tiket > 0:
    print(f"Sisa tiket: {jumlah_tiket}. Penjualan sedang berlangsung.")
    jumlah_tiket -= 1

print("Tiket habis!")

Sisa tiket: 5. Penjualan sedang berlangsung.
Sisa tiket: 4. Penjualan sedang berlangsung.
Sisa tiket: 3. Penjualan sedang berlangsung.
Sisa tiket: 2. Penjualan sedang berlangsung.
Sisa tiket: 1. Penjualan sedang berlangsung.
Tiket habis!


In [25]:
x = 1

while x <= 5:
    print("Nilai x =", x)
    x += 1 

Nilai x = 1
Nilai x = 2
Nilai x = 3
Nilai x = 4
Nilai x = 5


<div style="background-color:#f5f5f5; border-left:5px solid #FF0000; padding:12px;">
<b>Bahaya Hati Hati Perulangan Tak Terbatas (Infinite Loop) ⚠️<br></b>
Jika kondisi perulangan while Anda selalu bernilai True, perulangan tidak akan pernah berhenti. Ini akan membuat program Anda macet dan menghabiskan sumber daya CPU.
</div>

<div style="background-color:#f5f5f5; border-left:5px solid #555; padding:12px;">
Mengendalikan Alur Perulangan
Terkadang, Anda perlu mengubah alur perulangan dari dalam. Di sinilah break dan continue berperan.
<br>
<b>break</b>: Berfungsi sebagai "tombol keluar". Ini akan menghentikan seluruh perulangan secara total, bahkan jika kondisi while masih True atau masih ada elemen di for loop.
<br>
<b>continue</b>: Berfungsi sebagai "tombol lewati". Ini akan menghentikan iterasi saat ini dan langsung melompat ke iterasi berikutnya.

In [9]:
for angka in range(1, 10):
    if angka == 3:
        continue
    if angka == 7:
        break
    print(f"Angka: {angka}")

Angka: 1
Angka: 2
Angka: 4
Angka: 5
Angka: 6


### Konsep Perulangan Tingkat Lanjut

<div style="background-color:#f5f5f5; border-left:5px solid #555; padding:12px;">
1. Perulangan Bersarang (Nested Loops)
Ini adalah perulangan di dalam perulangan. Perulangan dalam akan dieksekusi sepenuhnya untuk setiap iterasi dari perulangan luar. Konsep ini sangat berguna untuk mengolah data dalam bentuk tabel atau matriks.
</div>

In [10]:
for baris in range(2):
    for kolom in range(3):
        print(f"({baris}, {kolom})", end=" ")
    print()

(0, 0) (0, 1) (0, 2) 
(1, 0) (1, 1) (1, 2) 


<div style="background-color:#f5f5f5; border-left:5px solid #555; padding:12px;">
2. Blok else dalam Perulangan
Ini adalah fitur unik Python. Blok else pada perulangan for atau while akan dieksekusi hanya jika perulangan selesai secara normal, artinya tidak dihentikan oleh break.
</div>

In [11]:
item_yang_dicari = "laptop"
inventaris = ["mouse", "keyboard", "monitor"]

for item in inventaris:
    if item == item_yang_dicari:
        print("Item ditemukan!")
        break
else:
    print("Item tidak ada di inventaris.")

Item tidak ada di inventaris.


<div style="background-color:#f5f5f5; border-left:5px solid #555; padding:12px;">
3. List (daftar)
list itu kumpulan data yang bisa diubah (mutable). isinya bisa campur-campur, angka, string, bool, dll.
</div>

In [27]:
buah = ["apel", "mangga", "jeruk"]

print(buah[0])   
print(buah[1])

# ubah data
buah[1] = "pisang"
print(buah)

# nambah data
buah.append("anggur")
print(buah)

# hapus data
buah.remove("apel")
print(buah)

apel
mangga
['apel', 'pisang', 'jeruk']
['apel', 'pisang', 'jeruk', 'anggur']
['pisang', 'jeruk', 'anggur']


<div style="background-color:#f5f5f5; border-left:5px solid #555; padding:12px;">
4. Dictionary (kamus)
dictionary itu pasangan key sama value, mirip tabel/objek.
</div>

In [28]:
orang = {
    "nama": "Budi",
    "umur": 29,
    "hobi": "gaming"
}

print(orang["nama"]) 
print(orang["umur"]) 

# ubah data
orang["umur"] = 20
print(orang)

# nambah key-value baru
orang["alamat"] = "Surabaya"
print(orang)

# hapus key-value
del orang["hobi"]
print(orang)


Budi
29
{'nama': 'Budi', 'umur': 20, 'hobi': 'gaming'}
{'nama': 'Budi', 'umur': 20, 'hobi': 'gaming', 'alamat': 'Surabaya'}
{'nama': 'Budi', 'umur': 20, 'alamat': 'Surabaya'}


In [40]:
buah = {"apel", "jeruk", "apel", "mangga"}

print(buah)

# nambah data
buah.add("pisang")
print(buah)

# hapus data
buah.remove("jeruk")
print(buah)

# operasi himpunan
a = {1, 2, 3}
b = {3, 4, 5}

print(a | b)  # gabungan -> {1,2,3,4,5}
print(a & b)  # irisan -> {3}
print(a - b)  # dikurangi -> {1,2}
print(b - a)  # dikurangi -> {4,5}

{'apel', 'jeruk', 'mangga'}
{'pisang', 'apel', 'jeruk', 'mangga'}
{'pisang', 'apel', 'mangga'}
{1, 2, 3, 4, 5}
{3}
{1, 2}
{4, 5}


<div style="
    background-color:#ffffff;
    border-bottom:8px solid green;
    box-shadow: 0 2px 6px rgba(0,0,0,0.15);
    border-radius:10px;
    padding:16px;
    margin:12px 0;
    font-family: 'Segoe UI', sans-serif;
">
<b>Apa perbedaan utama antara perulangan for dan while?</b>
<br>
<div style="margin-top:10px; background-color:#f9fafb; border-radius:6px; padding:8px 12px; border:1px dashed #3b82f6;">
✅ <i>Perulangan for digunakan saat jumlah iterasi sudah diketahui (misalnya, mengulang semua item dalam daftar), sedangkan perulangan while digunakan saat jumlah iterasi tidak pasti dan bergantung pada kondisi yang akan menjadi False di suatu titik.

</i>
</div>


<div style="
    background-color:#ffffff;
    border-bottom:8px solid green;
    box-shadow: 0 2px 6px rgba(0,0,0,0.15);
    border-radius:10px;
    padding:16px;
    margin:12px 0;
    font-family: 'Segoe UI', sans-serif;
">
<b>Apa fungsi dari pernyataan continue dan break di dalam sebuah perulangan?</b>
<br>
<div style="margin-top:10px; background-color:#f9fafb; border-radius:6px; padding:8px 12px; border:1px dashed #3b82f6;">
✅ <i>Pernyataan continue digunakan untuk melewati sisa kode pada iterasi saat ini dan langsung melanjutkan ke iterasi berikutnya. Sementara itu, break digunakan untuk menghentikan seluruh perulangan secara paksa.

</i>
</div>


## ⚜️ Strings Dan List

<div style="background-color:#f5f5f5; border-left:5px solid #555; padding:12px;">
<b>1. String</b>

String adalah tipe data yang berisi sekumpulan karakter (huruf, angka, simbol) yang ditulis di antara tanda kutip (' ' atau " ").

String bisa diakses per karakter menggunakan index, di mana index dimulai dari 0.
Kita juga bisa menelusuri karakter dalam string 1 per 1 menggunakan loop.
</div>

In [1]:
teks = "Python"

for huruf in teks:
    print(huruf)

P
y
t
h
o
n


<div style="background-color:#f5f5f5; border-left:5px solid #ff4444; padding:12px;">
   Sifat Penting: String adalah Immutable, Immutable berarti tidak dapat diubah setelah dibuat.
Bukan artinya string itu tidak bisa diganti sama sekali, tetapi kita tidak bisa mengubah sebagian karakter secara langsung, melainkan harus membuat string baru.
</div>

In [6]:
kata = "Halo"
kata = "J" + kata[1:]
print(kata)

Jalo


<div style="background-color:#f5f5f5; border-left:5px solid #555; padding:12px;">
    String sebagai Urutan (Sequence)
    String dianggap sebagai sequence atau rangkaian karakter.
    Setiap karakter memiliki posisi (index) dimulai dari 0 di sebelah kiri, dan bisa juga menggunakan index negatif dimulai dari -1 dari kanan.
    <br>
    Kapan String digunakan? String cocok digunakan ketika:
    <br>
Mengolah nama, alamat, atau teks input pengguna.
    <br>
Menampilkan informasi ke layar.
    <br>
Membentuk pesan, paragraf, dialog, dsb.
    <br>
Selain itu, string sangat sering dipakai di hampir semua program, karena komunikasi komputer ke pengguna biasanya berupa teks.
</div>


In [12]:
## membalik string
s1 = "abcdefgh"
print(s1[::-1])

## menampilkan tiap 2 huruf
s2 = "1234567890"
for i in range(0,len(s2),2):
    print(s2[i:i+2])

## membalik tiap kata
s3 = "hello world python"
kata_list = s3.split()
kata_terbalik = [k[::-1] for k in kata_list]
print(" ".join(kata_terbalik))

## membuat vokal menjadi huruf besar
s4 = "openai is cool"
vokal = "aiueo"
s4_new = "".join([c.upper() if c in vokal else c for c in s4])
print(s4_new)

## menampilkan kata index ganjil
s5 = "one two three four five six"
kata_list5 = s5.split()
for i in range(len(kata_list5)):
    if i % 2 == 1:
        print(kata_list5[i])

## menampilkan tiap huruf
s6 = "python"
for huruf in s6:
    print(huruf)

## ganti spasi jadi underscore
s7 = "machine learning"
print(s7.replace(" ","_"))

## slicing ambil 3 huruf pertama
s8 = "abcde"
print(s8[:3])

## slicing ambil 2 huruf terakhir
s9 = "12345"
print(s9[-2:])

## print tiap huruf dipisah spasi
s10 = "data"
print(" ".join(s10))

## reverse tiap kata
s11 = "good morning everyone"
kata_list11 = s11.split()
kata_rev11 = [k[::-1] for k in kata_list11]
print(" ".join(kata_rev11))

## bikin huruf pertama tiap kata jadi besar
s12 = "hello world python"
print(" ".join([k.capitalize() for k in s12.split()]))

## ulang string 3 kali
s13 = "abc"
print(s13*3)

## ambil index genap
s14 = "abcdefgh"
print(s14[::2])

## ambil index ganjil
s15 = "abcdefgh"
print(s15[1::2])

## upper semua huruf
s16 = "python is fun"
print(s16.upper())

## lower semua huruf
s17 = "PYTHON IS FUN"
print(s17.lower())

## ganti huruf tertentu
s18 = "banana"
print(s18.replace("a","o"))

## cek apakah string diawali huruf
s19 = "python"
print(s19.startswith("p"))

## cek apakah string diakhiri huruf
s20 = "python"
print(s20.endswith("n"))


hgfedcba
12
34
56
78
90
olleh dlrow nohtyp
OpEnAI Is cOOl
two
four
six
p
y
t
h
o
n
machine_learning
abc
45
d a t a
doog gninrom enoyreve
Hello World Python
abcabcabc
aceg
bdfh
PYTHON IS FUN
python is fun
bonono
True
True


<div style="background-color:#f5f5f5; border-left:5px solid #555; padding:12px;">
    “String comprehension” sebenarnya bukan fitur resmi Python seperti list comprehension, tetapi merupakan pola penggunaan list comprehension (atau comprehension lain) untuk mengolah string lalu menghasilkan string baru menggunakan ''.join(...). Karena string itu immutable (nggak bisa diubah langsung), maka Python memberi cara efisien untuk membuat string baru dari proses iterasi dan filter. Dengan teknik ini, kamu bisa memfilter huruf tertentu, memodifikasi huruf, menghapus karakter, sampai melakukan transformasi kompleks dalam 1–2 baris kode saja. Teknik ini jauh lebih cepat daripada melakukan loop manual yang menggabungkan string satu per satu.

Dalam praktiknya, string comprehension sering digunakan untuk hal-hal seperti menghapus spasi, mengekstrak huruf tertentu, mengubah format string, memproses input pengguna, atau mengamankan data (misal menghapus simbol berbahaya). Selain itu, cara ini juga membuat kode lebih bersih karena rumus logika berada dalam satu baris yang mudah dipahami. Python sendiri sangat mendukung pola ini karena join() sangat efisien untuk membentuk string besar tanpa overhead berlebihan.

FYI: Semua comprehension yang menghasilkan string itu bekerja dengan cara list comprehension → lalu digabung dengan "".join(...).

Dengan memahami teknik ini, kamu bisa membuat fungsi pemrosesan teks yang sangat cepat dan ringkas. Bahkan, jika kamu bekerja dengan algoritma text mining, tokenizing, data cleaning, dan NLP dasar, string comprehension adalah teknik wajib yang sering dipakai dalam data preparation.
</div>

In [17]:
text = "a p a   k a b a r "
hasil = "".join([c for c in text if c != " "])
print(hasil)

text = "HeLLo WoRLD"
hasil = "".join([c for c in text if c.isupper()])
print(hasil)

text = "Nomor: 0812-345-678"
hasil = "".join([c for c in text if c.isdigit()])
print(hasil)


apakabar
HLLWRLD
0812345678


<div style="background-color:#f5f5f5; border-left:5px solid #555; padding:12px;">
<b>2. LIST</b>

List adalah struktur data yang digunakan untuk menyimpan lebih dari satu nilai dalam satu variabel.
Bisa berupa data sejenis atau berbeda jenis.
</div>

In [7]:
["apel", "jeruk", "mangga"]
[10, 20, 30, 40]
["apel", 10, True, 3.14]

['apel', 10, True, 3.14]

<div style="background-color:#f5f5f5; border-left:5px solid #555; padding:12px;">
List sebagai Koleksi / Wadah

Kalau string tadi adalah rangkaian karakter, maka list adalah rangkaian elemen (items) yang bisa berupa apa saja.

List ibarat wadah atau rak penyimpanan:

Rak memiliki banyak slot

Setiap slot menyimpan satu barang

Kita bisa menambah, menghapus, atau mengganti barang kapan saja

Sifat Penting: List adalah Mutable

Mutable berarti elemen di dalam list bisa diubah tanpa harus membuat list baru.
</div>

In [8]:
buah = ["apel", "mangga", "jeruk"]
buah[1] = "pisang"

<div style="background-color:#f5f5f5; border-left:5px solid #555; padding:12px;">
<b>List juga bisa bertambah dan berkurang secara fleksibel:</b>

.append() → menambahkan di akhir

.remove() → menghapus berdasarkan isi

.pop() → menghapus berdasarkan index

Kapan List digunakan?

<b>List cocok ketika:</b>

Kamu ingin menyimpan kumpulan data

Data itu bisa berubah-ubah

Kamu ingin mengakses elemen berdasarkan urutan

Contoh penggunaannya banyak:

Menyimpan daftar siswa

Menyimpan data nilai

Menyimpan barang belanja

Menyimpan history tindakan pengguna
</div>

In [13]:
## reverse list
l1 = [1,2,3,4,5]
print(l1[::-1])

## ambil tiap 2 elemen
l2 = ["a","b","c","d","e","f"]
for i in range(0,len(l2),2):
    print(l2[i:i+2])

## gabungkan dua list
l3 = [1,2,3]
l4 = [4,5,6]
print(l3+l4)

## tambah elemen di akhir
l5 = ["apel","mangga"]
l5.append("jeruk")
print(l5)

## hapus elemen
l6 = [10,20,30,40]
l6.remove(20)
print(l6)

## ambil index genap
l7 = [0,1,2,3,4,5]
print(l7[::2])

## ambil index ganjil
l8 = [0,1,2,3,4,5]
print(l8[1::2])

## reverse tiap elemen string
l9 = ["abc","def","ghi"]
print([x[::-1] for x in l9])

## multiply list
l10 = [1,2,3]
print(l10*3)


[5, 4, 3, 2, 1]
['a', 'b']
['c', 'd']
['e', 'f']
[1, 2, 3, 4, 5, 6]
['apel', 'mangga', 'jeruk']
[10, 30, 40]
[0, 2, 4]
[1, 3, 5]
['cba', 'fed', 'ihg']
[1, 2, 3, 1, 2, 3, 1, 2, 3]


<div style="
    background-color:#ffffff;
    border-bottom:8px solid green;
    box-shadow: 0 2px 6px rgba(0,0,0,0.15);
    border-radius:10px;
    padding:16px;
    margin:12px 0;
    font-family: 'Segoe UI', sans-serif;
">
<b>Jelaskan dengan kata-katamu sendiri apa maksud dari string adalah immutable.</b>
<br>
<div style="margin-top:10px; background-color:#f9fafb; border-radius:6px; padding:8px 12px; border:1px dashed #3b82f6;">
✅ <i>String tidak bisa diubah sebagian setelah dibuat. Artinya, jika kita ingin mengganti salah satu huruf di dalam string, kita tidak bisa langsung mengeditnya, tetapi harus membuat string baru yang berisi perubahan tersebut.

Singkatnya: String tidak bisa dimodifikasi langsung.
</i>
</div>


<div style="
    background-color:#ffffff;
    border-bottom:8px solid green;
    box-shadow: 0 2px 6px rgba(0,0,0,0.15);
    border-radius:10px;
    padding:16px;
    margin:12px 0;
    font-family: 'Segoe UI', sans-serif;
">
<b>Jika string hanya menyimpan karakter, mengapa kita tetap butuh list di dalam pemrograman?</b>
<br>
<div style="margin-top:10px; background-color:#f9fafb; border-radius:6px; padding:8px 12px; border:1px dashed #3b82f6;">
✅ <i>Karena list dapat menyimpan berbagai macam data, bukan hanya karakter.
List juga digunakan ketika kita membutuhkan WADAH yang bisa:

berisi banyak nilai,

fleksibel diubah,

menambah/mengurangi elemen,

dan cocok untuk pengolahan data yang kompleks.

Dengan kata lain, string hanya untuk teks, sedangkan list digunakan untuk kumpulan data umum.
</i>
</div>


## ⚜️ Function

<div style="background-color:#f5f5f5; border-left:5px solid #555; padding:12px;">
    Function adalah blok kode terorganisir yang dibuat untuk menjalankan tugas tertentu secara berulang tanpa harus menulis ulang kodenya. Python menggunakan konsep modular programming, yaitu memecah program besar menjadi bagian-bagian kecil yang mudah dikelola. Fungsi membantu meningkatkan keterbacaan, mengurangi duplikasi kode, dan mempermudah perbaikan karena kita cukup mengubah logika di satu tempat saja. Function juga dapat menerima parameter sebagai input dan mengembalikan nilai menggunakan return, sehingga bisa digunakan sebagai komponen dalam perhitungan yang lebih besar. Dengan adanya function, kita bisa membuat kode yang lebih “reusable” dan scalable.
</div>

In [2]:
def sapa(nama):
    return f"Halo, {nama}!"
print(sapa("Budi"))

def salam():
    return "Selamat pagi semuanya!"
print(salam())

def tambah(a, b):
    return a + b
print(tambah(5, 7))

def greet(nama="Tamu"):
    return f"Welcome, {nama}!"
print(greet())
print(greet("Rani"))

def luas_persegi(sisi):
    return sisi * sisi
print(luas_persegi(6))

def cetak_data(data):
    print("Isi data:", data)
cetak_data("mantap")

def cek_umur(umur):
    if umur >= 18:
        return "Dewasa"
    return "Anak-anak"
print(cek_umur(20))

def ulang_kata(kata, n):
    for i in range(n):
        print(kata)
ulang_kata("Hallo", 3)

def buat_list(n):
    hasil = []
    for i in range(1, n+1):
        hasil.append(i)
    return hasil
print(buat_list(5))

def jumlah_semua(*angka):
    return sum(angka)
print(jumlah_semua(1, 2, 3, 4, 5))

def info(**data):
    return data
print(info(nama="Eka", umur=20))

def operasi(a, b):
    return a + b, a - b, a * b, a / b
print(operasi(10, 2))

def is_genap(n):
    return n % 2 == 0
print(is_genap(6))

def faktorial(n):
    hasil = 1
    for i in range(1, n+1):
        hasil *= i
    return hasil
print(faktorial(5))

def balik_kata(kata):
    return kata[::-1]
print(balik_kata("python"))

def hitung_vokal(teks):
    vokal = "aiueoAIUEO"
    total = 0
    for huruf in teks:
        if huruf in vokal:
            total += 1
    return total
print(hitung_vokal("Indonesia"))

def max_dari_tiga(a, b, c):
    return max(a, b, c)
print(max_dari_tiga(10, 3, 7))

def tampilkan_range(a, b):
    for i in range(a, b+1):
        print(i)
tampilkan_range(3, 7)

def pangkat(angka, pangkat):
    return angka ** pangkat
print(pangkat(2, 3))

def konversi_celcius_ke_f(c):
    return (c * 9/5) + 32
print(konversi_celcius_ke_f(30))

def is_palindrome(teks):
    return teks == teks[::-1]
print(is_palindrome("level"))


Halo, Budi!
Selamat pagi semuanya!
12
Welcome, Tamu!
Welcome, Rani!
36
Isi data: mantap
Dewasa
Hallo
Hallo
Hallo
[1, 2, 3, 4, 5]
15
{'nama': 'Eka', 'umur': 20}
(12, 8, 20, 5.0)
True
120
nohtyp
5
10
3
4
5
6
7
8
86.0
True


<div style="background-color:#f5f5f5; border-left:5px solid #555; padding:12px;">
    FYI: Python memperlakukan function sebagai “first-class object,” artinya fungsi bisa disimpan dalam variabel, dijadikan parameter, bahkan dikembalikan dari fungsi lain.
</div>

<div style="background-color:#f5f5f5; border-left:5px solid #555; padding:12px;">
    Secara konsep, function juga penting karena mendorong gaya penulisan kode yang lebih rapih dan modular. Dalam proyek besar, pengembang biasanya membuat puluhan hingga ratusan fungsi yang masing-masing menangani bagian kecil dari program, sehingga kolaborasi antar programmer menjadi lebih mudah. Function juga meningkatkan skalabilitas aplikasi: cukup mengubah isi function, dan semua bagian program yang menggunakannya akan otomatis ikut berubah. Inilah alasan kenapa function adalah fondasi dasar dalam hampir semua bahasa pemrograman modern, termasuk Python.
</div>

In [3]:
def kali(a, b):
    return a * b
print(kali(4, 8))

def bagi(a, b):
    return a / b
print(bagi(20, 5))

def sapa_panjang(nama):
    return "Halo, selamat datang " + nama
print(sapa_panjang("Rizal"))

def rata_rata(a, b, c):
    return (a + b + c) / 3
print(rata_rata(10, 20, 40))

def gabung_kata(a, b):
    return a + " " + b
print(gabung_kata("Python", "Keren"))

def hitung_huruf(teks):
    return len(teks)
print(hitung_huruf("Selamat"))

def ganjil_genap(n):
    if n % 2 == 0:
        return "Genap"
    return "Ganjil"
print(ganjil_genap(13))

def perkalian_berulang(n):
    total = 1
    for i in range(1, n+1):
        total *= i
    return total
print(perkalian_berulang(6))

def list_ganjil(n):
    hasil = []
    for i in range(1, n+1, 2):
        hasil.append(i)
    return hasil
print(list_ganjil(15))

def huruf_besar(teks):
    return teks.upper()
print(huruf_besar("kodIng"))

def huruf_kecil(teks):
    return teks.lower()
print(huruf_kecil("KODING"))

def selisih(a, b):
    return abs(a - b)
print(selisih(25, 40))

def jumlah_list(data):
    return sum(data)
print(jumlah_list([5, 10, 15, 20]))

def ambil_depan(teks, jumlah):
    return teks[:jumlah]
print(ambil_depan("PROGRAMMER", 5))

def ambil_belakang(teks, jumlah):
    return teks[-jumlah:]
print(ambil_belakang("PROGRAMMER", 4))

def terbalik_list(data):
    return data[::-1]
print(terbalik_list([1, 2, 3, 4, 5]))

def average_list(data):
    return sum(data) / len(data)
print(average_list([2, 4, 6, 8]))

def cek_pos_neg(n):
    if n > 0:
        return "Positif"
    elif n < 0:
        return "Negatif"
    return "Nol"
print(cek_pos_neg(-5))

def double_kata(teks):
    return teks * 2
print(double_kata("Hi "))

def list_kelipatan(n):
    hasil = []
    for i in range(1, 6):
        hasil.append(n * i)
    return hasil
print(list_kelipatan(7))


32
4.0
Halo, selamat datang Rizal
23.333333333333332
Python Keren
7
Ganjil
720
[1, 3, 5, 7, 9, 11, 13, 15]
KODING
koding
15
50
PROGR
MMER
[5, 4, 3, 2, 1]
5.0
Negatif
Hi Hi 
[7, 14, 21, 28, 35]


<div style="
    background-color:#ffffff;
    border-bottom:8px solid green;
    box-shadow: 0 2px 6px rgba(0,0,0,0.15);
    border-radius:10px;
    padding:16px;
    margin:12px 0;
    font-family: 'Segoe UI', sans-serif;
">
<b>Buat function luas_persegi yang menerima 1 parameter (sisi) dan mengembalikan luasnya.</b>
<br>
<div style="margin-top:10px; background-color:#f9fafb; border-radius:6px; padding:8px 12px; border:1px dashed #3b82f6;">
✅ <i>def luas_persegi(s):
    return s * s

</i>
</div>


## ⚜️ Recursive Function

<div style="background-color:#f5f5f5; border-left:5px solid #555; padding:12px;">
    Recursive function adalah fungsi yang memanggil dirinya sendiri sampai mencapai kondisi penghentian (base case). Pendekatan rekursif digunakan untuk menyelesaikan masalah yang dapat dipecah menjadi sub-masalah yang struktur penyelesaiannya mirip, seperti faktorial, fibonacci, tree traversal, dan divide-and-conquer. Rekursi sangat elegan tetapi harus ditulis dengan hati-hati agar tidak memicu infinite recursion atau stack overflow. Python memiliki batas rekursi tertentu, sehingga rekursi dalam Python lebih cocok untuk masalah yang tidak terlalu dalam kedalamannya.
</div>

<div style="background-color:#f5f5f5; border-left:5px solid #555; padding:12px;">
    FYI: Python punya batas kedalaman rekursi bawaan (biasanya 1000 pemanggilan), supaya program tidak crash akibat pemanggilan fungsi yang tak berujung.
</div>

In [5]:
def faktorial(n):
    if n == 1:
        return 1
    return n * faktorial(n-1)
faktorial(5)

120

<div style="background-color:#f5f5f5; border-left:5px solid #555; padding:12px;">
    Meski rekursi terlihat elegan dan bersih, penggunaannya harus hati-hati. Tanpa base case yang jelas, fungsi dapat terjebak dalam lingkaran pemanggilan tak terbatas atau menghabiskan banyak memori stack. Untuk beberapa kasus, pendekatan iteratif (loop) lebih efisien dan cepat. Namun, rekursi tetap menjadi teknik penting dalam pemrograman karena cocok untuk masalah yang memiliki pola pecahan (divide-and-conquer), seperti quicksort, mergesort, dan pencarian pada struktur data yang bercabang. Rekursi bukan hanya teknik, tapi juga cara berpikir dalam memecah masalah.
</div>

In [5]:
def hitung_mundur(n):
    if n == 0:
        return
    print(n)
    hitung_mundur(n-1)

def fibonacci(n):
    if n <= 1:
        return n
    return fibonacci(n-1) + fibonacci(n-2)

def cari(data, target, index=0):
    if index == len(data):
        return False
    if data[index] == target:
        return True
    return cari(data, target, index + 1)

def balik(kata):
    if len(kata) <= 1:
        return kata
    return balik(kata[1:]) + kata[0]

hitung_mundur(5)
print(fibonacci(6))
print(cari([3, 6, 9, 12], 9))
print(balik("kucing"))

5
4
3
2
1
8
True
gnicuk


In [6]:
def jumlah_digit(n):
    if n == 0:
        return 0
    return (n % 10) + jumlah_digit(n // 10)
print(jumlah_digit(12345))

def pangkat_rekursif(a, b):
    if b == 0:
        return 1
    return a * pangkat_rekursif(a, b - 1)
print(pangkat_rekursif(3, 4))

def hitung_huruf(teks):
    if teks == "":
        return 0
    return 1 + hitung_huruf(teks[1:])
print(hitung_huruf("python"))

def sum_list(data, i=0):
    if i == len(data):
        return 0
    return data[i] + sum_list(data, i + 1)
print(sum_list([3, 6, 9, 12]))

def minimum_list(data, i=0):
    if i == len(data) - 1:
        return data[i]
    nilai_min = minimum_list(data, i + 1)
    return data[i] if data[i] < nilai_min else nilai_min
print(minimum_list([7, 3, 8, 2, 9]))

def kelipatan_rekursif(n, k=1, hasil=None):
    if hasil is None:
        hasil = []
    if k > 5:
        return hasil
    hasil.append(n * k)
    return kelipatan_rekursif(n, k + 1, hasil)
print(kelipatan_rekursif(4))

def delete_char(teks, char):
    if teks == "":
        return ""
    sisa = delete_char(teks[1:], char)
    if teks[0] == char:
        return sisa
    return teks[0] + sisa
print(delete_char("banana", "a"))

def faktorial_genap(n):
    if n <= 0:
        return 1
    return n * faktorial_genap(n - 2)
print(faktorial_genap(8))


15
81
6
30
2
[4, 8, 12, 16, 20]
bnn
384


<div style="
    background-color:#ffffff;
    border-bottom:8px solid green;
    box-shadow: 0 2px 6px rgba(0,0,0,0.15);
    border-radius:10px;
    padding:16px;
    margin:12px 0;
    font-family: 'Segoe UI', sans-serif;
">
<b>Buat fungsi rekursif bernama jumlah(n) yang menjumlahkan angka 1 sampai n.</b>
<br>
<div style="margin-top:10px; background-color:#f9fafb; border-radius:6px; padding:8px 12px; border:1px dashed #3b82f6;">
✅ <i>def jumlah(n):
    if n == 1:
        return 1
    return n + jumlah(n-1)
</i>
</div>


## ⚜️ List Comprehension & Tuples

<div style="background-color:#f5f5f5; border-left:5px solid #555; padding:12px;">
    List comprehension adalah cara ringkas untuk membuat list baru dengan melakukan iterasi (looping) dan operasi tertentu hanya dalam satu baris kode. Konsep ini menggabungkan kekuatan loop dan kondisi menjadi sintaks yang bersih dan cepat. Python menggunakan teknik ini untuk meningkatkan efisiensi dan keterbacaan, terutama dalam operasi transformasi data, filtering, dan mapping. Dibandingkan loop biasa, list comprehension biasanya lebih cepat karena dioptimasi pada level internal interpreter Python.
</div>

In [3]:
angka = [x*2 for x in range(5)]
print(angka)

[0, 2, 4, 6, 8]


<div style="background-color:#f5f5f5; border-left:5px solid #555; padding:12px;">
    FYI: List comprehension bisa digabung dengan if, bahkan if–else, untuk membuat list yang kompleks.

Dari sudut pandang pemrograman fungsional, list comprehension mencerminkan konsep mapping dan filtering. Ketika data besar harus diproses, list comprehension membuat kode lebih ringkas sehingga programmer bisa fokus pada logika, bukan struktur loop. Meski begitu, penggunaan berlebihan—misalnya membuat comprehension yang berlapis-lapis dan terlalu panjang—justru membuat kode sulit dibaca. Jadi, list comprehension sebaiknya digunakan ketika transformasi datanya masih cukup sederhana dan jelas bagi pembaca.
</div>

In [10]:
angka = [x for x in range(1, 11)]
kata = "HeLLo WoRLD"
kapital = [c for c in kata if c.isupper()]
nilai_tinggi = [
    n
    for data in siswa.values()
    for n in data["nilai"]
    if n > 90
]

print(angka)
print(kata)
print(kapital)
print(nilai_tinggi)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
HeLLo WoRLD
['H', 'L', 'L', 'W', 'R', 'L', 'D']
[92, 94, 97, 93, 91]


In [7]:
angka_kuadrat = [x*x for x in range(1, 6)]
print(angka_kuadrat)

angka_genap = [num for num in range(1, 21) if num % 2 == 0]
print(angka_genap)

huruf_besar = [char.upper() for char in "programmer"]
print(huruf_besar)

panjang_kata = [len(kata) for kata in ["kuda", "komputer", "meja"]]
print(panjang_kata)

angka_positif = [n for n in [-3, -1, 0, 2, 5] if n > 0]
print(angka_positif)

angka_kelipatan_tiga = [n for n in range(1, 31) if n % 3 == 0]
print(angka_kelipatan_tiga)

kata_pendek = [word for word in ["api", "batu", "program", "gelas"] if len(word) <= 4]
print(kata_pendek)

huruf_pertama = [nama[0] for nama in ["Andi", "Beni", "Cici"]]
print(huruf_pertama)

perkalian_lima = [n*5 for n in range(1, 6)]
print(perkalian_lima)

angka_ganjil = [n for n in range(1, 15) if n % 2 != 0]
print(angka_ganjil)

kata_kapital = [w.capitalize() for w in ["merah", "kuning", "hijau"]]
print(kata_kapital)

nilai_pangkat_tiga = [n**3 for n in range(1, 5)]
print(nilai_pangkat_tiga)

angka_diatas_tujuh = [n for n in [2, 7, 8, 10, 3] if n > 7]
print(angka_diatas_tujuh)

angka_mutlak = [abs(n) for n in [-4, 0, 7, -9]]
print(angka_mutlak)

kata_tanpa_spasi = [kata.strip() for kata in [" ayam ", " bebek ", " sapi "]]
print(kata_tanpa_spasi)

panjang_huruf = [len(h) for h in "pertanian"]
print(panjang_huruf)

dua_dimensi = [[n for n in range(3)] for _ in range(3)]
print(dua_dimensi)

list_boolean = [x > 5 for x in range(1, 11)]
print(list_boolean)

angka_dikali_dua = [n*2 for n in [3, 6, 9, 12]]
print(angka_dikali_dua)

huruf_vokal = [h for h in "universitas" if h in "aeiou"]
print(huruf_vokal)

[1, 4, 9, 16, 25]
[2, 4, 6, 8, 10, 12, 14, 16, 18, 20]
['P', 'R', 'O', 'G', 'R', 'A', 'M', 'M', 'E', 'R']
[4, 8, 4]
[2, 5]
[3, 6, 9, 12, 15, 18, 21, 24, 27, 30]
['api', 'batu']
['A', 'B', 'C']
[5, 10, 15, 20, 25]
[1, 3, 5, 7, 9, 11, 13]
['Merah', 'Kuning', 'Hijau']
[1, 8, 27, 64]
[8, 10]
[4, 0, 7, 9]
['ayam', 'bebek', 'sapi']
[1, 1, 1, 1, 1, 1, 1, 1, 1]
[[0, 1, 2], [0, 1, 2], [0, 1, 2]]
[False, False, False, False, False, True, True, True, True, True]
[6, 12, 18, 24]
['u', 'i', 'e', 'i', 'a']


<div style="background-color:#f5f5f5; border-left:5px solid #555; padding:12px;">
    <b>Tuple</b> adalah struktur data mirip list tetapi bersifat immutable—artinya nilainya tidak dapat diubah setelah dibuat. Karena bersifat tetap, tuple sering digunakan untuk data yang tidak boleh dimodifikasi, seperti koordinat, konfigurasi, atau hasil return yang berisi beberapa nilai. Keuntungan tuple adalah lebih aman, lebih cepat diproses, dan bisa digunakan sebagai key pada dictionary (karena hashable), tidak seperti list. Tuple juga dapat menyimpan berbagai jenis data, dan operasi dasarnya sama dengan list, hanya saja tanpa kemampuan untuk menambah atau menghapus item.
</div>

In [8]:
data = (10, 20)

print(data)

(10, 20)


<div style="background-color:#f5f5f5; border-left:5px solid #555; padding:12px;">
    Immutability membuat tuple juga bisa digunakan sebagai key dalam dictionary, sesuatu yang tidak bisa dilakukan oleh list. Kecepatan tuple juga membuatnya cocok untuk operasi yang membutuhkan performa tinggi. Dalam praktik, tuple sering dipakai ketika fungsi perlu mengembalikan lebih dari satu nilai, karena tuple secara otomatis dibuat ketika kita mengembalikan beberapa item sekaligus. Dengan sifatnya yang ringan dan stabil, tuple menjadi salah satu struktur data penting dalam Python untuk data yang sifatnya tetap.
</div>

In [14]:
siswa_tuple = ("Rafi", 17, "XI RPL")
nama, umur, kelas = siswa_tuple

def hitung(a, b):
    return a+b, a*b, a-b

tambah, kali, kurang = hitung(10, 5)

ranking = [
    ("Dewi", 95),
    ("Rafi", 88),
    ("Andi", 76),
]

ranking_sorted = sorted(ranking, key=lambda x: x[1], reverse=True)

print(siswa_tuple)
hitung(7,9)
print(tambah)

print(ranking_sorted)

('Rafi', 17, 'XI RPL')
15
[('Dewi', 95), ('Rafi', 88), ('Andi', 76)]


In [10]:
daftar_buah = ("apel", "jeruk", "mangga")
print(daftar_buah)

angka_gabungan_tuple = (1, 2, 3) + (4, 5, 6)
print(angka_gabungan_tuple)

kata_diulang_tuple = ("halo",) * 4
print(kata_diulang_tuple)

hewan_peliharaan = ("kucing", "anjing", "ular")
hewan_peliharaan = hewan_peliharaan + ("burung",)
print(hewan_peliharaan)

huruf_awal_list = list(("a", "b", "c"))
huruf_awal_list.append("d")
huruf_awal_tuple = tuple(huruf_awal_list)
print(huruf_awal_tuple)

warna_subset_tuple = ("merah", "hijau", "biru", "kuning")[1:3]
print(warna_subset_tuple)

profil_pengguna_tuple = ("Andi", 18, "Bangkalan")
nama_pengguna, umur_pengguna, kota_pengguna = profil_pengguna_tuple
print(nama_pengguna, umur_pengguna, kota_pengguna)

warna_lampu_lalu_lintas = ("merah", "kuning", "hijau")
print(warna_lampu_lalu_lintas.index("kuning"))

jumlah_huruf_a = ("a", "b", "a", "a", "c").count("a")
print(jumlah_huruf_a)

pasangan_huruf_angka = (("A", 1), ("B", 2), ("C", 3))
print(pasangan_huruf_angka[2][1])

angka_urutan_tuple = tuple(range(10))
print(angka_urutan_tuple)

tuple_satu_item = ("python",)
print(len(tuple_satu_item))

daftar_hewan_ternak = ("ayam", "sapi", "kambing")
print("sapi" in daftar_hewan_ternak)

tuple_dengan_duplikat = ("apel", "apel", "apel")
print(set(tuple_dengan_duplikat))

tuple_huruf_makanan = ("m", "a", "k", "a", "n")
print("".join(tuple_huruf_makanan))

tuple_abjad_awal = ("a", "b")
tuple_abjad_akhir = ("c", "d")
tuple_abjad_lengkap = tuple_abjad_awal + tuple_abjad_akhir
print(tuple_abjad_lengkap)

kata_dalam_tuple = ("halo", "dunia", "python")
panjang_kata_tuple = [len(k) for k in kata_dalam_tuple]
print(panjang_kata_tuple)

nilai_angka_tuple = (10, 20, 30)
print(sum(nilai_angka_tuple))

reversed_kata_tuple = ("m", "e", "j", "a")[::-1]
print(reversed_kata_tuple)

kuadrat_angka_tuple = tuple(x * x for x in range(1, 6))
print(kuadrat_angka_tuple)

('apel', 'jeruk', 'mangga')
(1, 2, 3, 4, 5, 6)
('halo', 'halo', 'halo', 'halo')
('kucing', 'anjing', 'ular', 'burung')
('a', 'b', 'c', 'd')
('hijau', 'biru')
Andi 18 Bangkalan
1
3
3
(0, 1, 2, 3, 4, 5, 6, 7, 8, 9)
1
True
{'apel'}
makan
('a', 'b', 'c', 'd')
[4, 5, 6]
60
('a', 'j', 'e', 'm')
(1, 4, 9, 16, 25)


## ⚜️ Dictionary

<div style="background-color:#f5f5f5; border-left:5px solid #555; padding:12px;">
    Dictionary adalah struktur data berupa pasangan key-value yang memungkinkan penyimpanan dan pengambilan data secara cepat menggunakan key. Berbeda dengan list yang diakses melalui indeks, dictionary diakses melalui key unik yang dapat berupa string, number, atau tuple. Dictionary sangat berguna untuk menyimpan data terstruktur seperti objek, konfigurasi, mapping, atau data JSON. Di bawahnya, dictionary menggunakan struktur hash table, sehingga operasi pencarian dan penyisipan data rata-rata sangat cepat. Python modern juga menjaga insertion order, artinya urutan item yang dimasukkan tetap dipertahankan.
</div>

In [6]:
mhs = {"nama": "Ani", "umur": 19}

print(mhs)

{'nama': 'Ani', 'umur': 19}


<div style="background-color:#f5f5f5; border-left:5px solid #555; padding:12px;">
    Di balik layar, dictionary menggunakan struktur hash table, sehingga operasi seperti mencari, menambah, atau menghapus data menjadi sangat cepat bahkan ketika data berjumlah besar. Keunggulan lainnya adalah kemampuannya untuk menyimpan struktur kompleks, seperti list di dalam dictionary atau dictionary di dalam dictionary. Karena fleksibilitas dan kecepatan inilah dictionary menjadi struktur data yang paling sering digunakan dalam pengembangan aplikasi Python, terutama untuk pemrosesan data, backend API, dan manajemen objek.
</div>

In [9]:
siswa = {
    "001": {
        "nama": "Rafi",
        "kelas": "XI RPL",
        "nilai": [85, 90, 88],
        "alamat": "Bangkal",
        "aktif": True
    },
    "002": {
        "nama": "Dewi",
        "kelas": "XI TKJ",
        "nilai": [92, 89, 94],
        "alamat": "Arosbaya",
        "aktif": True
    },
    "003": {
        "nama": "Andi",
        "kelas": "X RPL",
        "nilai": [75, 80, 70],
        "alamat": "Burneh",
        "aktif": False
    }
}

siswa["004"] = {
    "nama": "Sinta",
    "kelas": "XI RPL",
    "nilai": [90, 93, 91],
    "alamat": "Galis",
    "aktif": True
}

del siswa["003"]
# atau: siswa.pop("003")

siswa["002"]["kelas"] = "XII TKJ"
siswa["002"]["nilai"].append(97)
siswa["002"]["alamat"] = "Kebun"

# Update sekaligus
siswa["001"].update({
    "kelas": "XII RPL",
    "aktif": False
})

if "002" in siswa:
    print("Data siswa ditemukan")

for nis, data in siswa.items():
    print(nis, "-", data["nama"], "-", data["kelas"])


Data siswa ditemukan
001 - Rafi - XII RPL
002 - Dewi - XII TKJ
004 - Sinta - XI RPL


In [11]:
informasi_pengguna = {"nama": "Rudi", "umur": 20}
print(informasi_pengguna)

stok_buah_toko = {"apel": 10}
stok_buah_toko["jeruk"] = 5
print(stok_buah_toko)

koordinat_titik = {"x": 1, "y": 2}
koordinat_titik["y"] = 10
print(koordinat_titik)

harga_makanan_warung = {"soto": 15000}
harga_makanan_warung.update({"bakso": 12000})
print(harga_makanan_warung)

data_login_admin = {"username": "admin", "password": "123"}
print(data_login_admin["username"])

kamus_huruf = {"huruf1": "A", "huruf2": "B"}
kamus_huruf.pop("huruf1")
print(kamus_huruf)

nilai_siswa_kelas = {"andi": 80, "budi": 90, "cici": 85}
print(list(nilai_siswa_kelas.keys()))

print(list(nilai_siswa_kelas.values()))

print(list(nilai_siswa_kelas.items()))

informasi_barang_toko = {"barang": "roti"}
informasi_barang_toko.setdefault("stok", 30)
print(informasi_barang_toko)

pengaturan_aplikasi = {"mode": "dark", "volume": 50}
print(pengaturan_aplikasi.get("mode"))

data_harga_buah = {"apel": 5, "jeruk": 3}
del data_harga_buah["jeruk"]
print(data_harga_buah)

nilai_angka_pemain = {"p1": 10, "p2": 20, "p3": 30}
for pemain, nilai in nilai_angka_pemain.items():
    print(pemain, nilai)

biodata_pendek = {"nama": "Doni", "umur": 19}
print("umur" in biodata_pendek)

harga_barang_diskon = {"harga": 15000}
harga_barang_diskon["harga"] += 5000
print(harga_barang_diskon)

stok_minuman = {"teh": 10, "kopi": 15}
stok_minuman["teh"] -= 3
print(stok_minuman)

data_mahasiswa_kelas = {"mhs": ["andi", "budi"]}
data_mahasiswa_kelas["mhs"].append("cici")
print(data_mahasiswa_kelas)

detail_buah_import = {"buah": ("apel", "jeruk")}
print(detail_buah_import["buah"][1])

konfigurasi_sistem = {"versi": 1.0, "mode": "auto"}
konfigurasi_sistem.clear()
print(konfigurasi_sistem)

nilai_uas_siswa = {"andi": 3, "budi": 5}
nilai_uas_dikali_dua = {k: v * 2 for k, v in nilai_uas_siswa.items()}
print(nilai_uas_dikali_dua)

informasi_peserta = {"nama": "Rahmat", "umur": 22}
print(list(informasi_peserta))

{'nama': 'Rudi', 'umur': 20}
{'apel': 10, 'jeruk': 5}
{'x': 1, 'y': 10}
{'soto': 15000, 'bakso': 12000}
admin
{'huruf2': 'B'}
['andi', 'budi', 'cici']
[80, 90, 85]
[('andi', 80), ('budi', 90), ('cici', 85)]
{'barang': 'roti', 'stok': 30}
dark
{'apel': 5}
p1 10
p2 20
p3 30
True
{'harga': 20000}
{'teh': 7, 'kopi': 15}
{'mhs': ['andi', 'budi', 'cici']}
jeruk
{}
{'andi': 6, 'budi': 10}
['nama', 'umur']


<div style="
    background-color:#ffffff;
    border-bottom:8px solid green;
    box-shadow: 0 2px 6px rgba(0,0,0,0.15);
    border-radius:10px;
    padding:16px;
    margin:12px 0;
    font-family: 'Segoe UI', sans-serif;
">
<b>Buat dictionary siswa dengan key: nama, kelas, nilai. Ambil nilai dari key kelas.</b>
<br>
<div style="margin-top:10px; background-color:#f9fafb; border-radius:6px; padding:8px 12px; border:1px dashed #3b82f6;">
✅ <i>
    siswa = {"nama": "Doni", "kelas": "XI RPL", "nilai": 90}
    kelas = siswa["kelas"]
</i>
</div>


## ⚜️Tugas Flowchart

### → Setup & Docs: [Documentation](https://mermaid.js.org/syntax/flowchart.html) (Library)

In [2]:
import importlib.util
if importlib.util.find_spec("mermaid_magic") is None:
    !pip install mermaid-magic
    print("Mermaid-magic Berhasil di instal.")
    print("Harap restart kernel")
else:
    print('----------------------------------')
    print('Mermaid Magic Berhasil Terhubung')
    print('----------------------------------')
    %load_ext mermaid_magic
    %reload_ext mermaid_magic

# %%mermaid

#---
# title: "Ini\nContoh Judul"
# ---
# flowchart TD
#     A[/Start/] --> B{APAKAH INI BEKERJA?}
#     B -->|Yes| C[Great!]
#     B -->|No| D[Debug]
#     D --> B
#     C --> E[/End/]

----------------------------------
Mermaid Magic Berhasil Terhubung
----------------------------------


### → Pengerjaan

#### Nomer 1 (Menghitung Luas bangun datar hasil input User)

In [11]:
import math

print("Pilih bentuk yang ingin dihitung luasnya:")
print("1. Bujur Sangkar")
print("2. Segitiga")
print("3. Lingkaran")

pilihan = input("Masukkan pilihan Anda (1/2/3): ")

if pilihan == '1':
  try:
    sisi = float(input("Masukkan panjang sisi bujur sangkar: "))
    if sisi < 0:
        print("Error: Sisi tidak boleh bernilai negatif.")
    else:
        luas = sisi * sisi
        print("Luas bujur sangkar adalah:", luas)
  except ValueError:
    print("Input tidak valid, harap masukkan sebuah angka.")

elif pilihan == '2':
  try:
    alas = float(input("Masukkan panjang alas segitiga: "))
    tinggi = float(input("Masukkan tinggi segitiga: "))
    if alas < 0 or tinggi < 0:
        print("Error: Alas dan tinggi tidak boleh bernilai negatif.")
    else:
        luas = 0.5 * alas * tinggi
        print("Luas segitiga adalah:", luas)
  except ValueError:
    print("Input tidak valid, harap masukkan sebuah angka.")

elif pilihan == '3':
  try:
    jari_jari = float(input("Masukkan panjang jari-jari lingkaran: "))
    if jari_jari < 0:
        print("Error: Jari-jari tidak boleh bernilai negatif.")
    else:
        luas = math.pi * (jari_jari ** 2)
        print("Luas lingkaran adalah:", luas)
  except ValueError:
    print("Input tidak valid, harap masukkan sebuah angka.")

else:
  print("Pilihan tidak valid. Anda harus memasukkan angka 1, 2, atau 3.")

Pilih bentuk yang ingin dihitung luasnya:
1. Bujur Sangkar
2. Segitiga
3. Lingkaran


Masukkan pilihan Anda (1/2/3):  1
Masukkan panjang sisi bujur sangkar:  4


Luas bujur sangkar adalah: 16.0


<img src="./no1.png" width=500/>

#### Nomer 2 (Bilangan Ganjil sebanyak n)

In [2]:
try:
    n = int(input("Masukkan jumlah bilangan ganjil yang ingin ditampilkan: "))

    if n <= 0:
        print("Harap masukkan bilangan bulat positif.")
    else:
        print(f"\n{n} bilangan ganjil pertama adalah:\n")
        for i in range(n):
            bilangan_ganjil = 2 * i + 1
            print(bilangan_ganjil)

except ValueError:
    print("Input tidak valid. Harap masukkan sebuah angka.")

Masukkan jumlah bilangan ganjil yang ingin ditampilkan:  4



4 bilangan ganjil pertama adalah:

1
3
5
7


<img src="./no2.png" width=300/>

#### Nomer 3 (Menghitung Deret yang sudah di tentukan)

In [15]:
try:
    total_target = float(input("Masukkan nilai total yang diinginkan: "))

    jumlah_sekarang = 0.0
    n = 1

    print("-" * 30)
    print("Suku-suku deret yang ditampilkan:")
    print("-" * 30)

    while jumlah_sekarang < total_target:
        pembilang = 3 ** n
        penyebut = (4 * n) + 10

        suku_sekarang = pembilang / penyebut
        jumlah_sekarang += suku_sekarang        
        print(f"Suku ke-{n}: {pembilang}/{penyebut} (Nilai: {suku_sekarang:.2f})")
        n += 1
    
    print("-" * 30)

except ValueError:
    print("Input tidak valid. Harap masukkan sebuah angka.")

Masukkan nilai total yang diinginkan:  34.6


------------------------------
Suku-suku deret yang ditampilkan:
------------------------------
Suku ke-1: 3/14 (Nilai: 0.21)
Suku ke-2: 9/18 (Nilai: 0.50)
Suku ke-3: 27/22 (Nilai: 1.23)
Suku ke-4: 81/26 (Nilai: 3.12)
Suku ke-5: 243/30 (Nilai: 8.10)
Suku ke-6: 729/34 (Nilai: 21.44)
Suku ke-7: 2187/38 (Nilai: 57.55)
------------------------------


<img src="./no3.png" width=300/>

#### Nomer 4 (Menghitung Deret yang sudah di tentukan #2)

In [17]:
try:
    n_suku = int(input("Masukkan jumlah suku yang ingin ditampilkan: "))

    if n_suku <= 0:
        print("Harap masukkan bilangan bulat positif.")
    else:
        print(f"Menampilkan {n_suku} suku deret:")
        penyebut = 14
        for i in range(n_suku):
            # i dimulai dari 0
            pembilang = i + 3
            
            if i % 2 == 0:
                print(f"Suku ke-{i+1}: {pembilang}/{penyebut}")
            else:
                print(f"Suku ke-{i+1}: -{pembilang}/{penyebut}")
            penyebut *= 2

except ValueError:
    print("Input tidak valid, harap masukkan angka.")

Masukkan jumlah suku yang ingin ditampilkan:  9


Menampilkan 9 suku deret:
Suku ke-1: 3/14
Suku ke-2: -4/28
Suku ke-3: 5/56
Suku ke-4: -6/112
Suku ke-5: 7/224
Suku ke-6: -8/448
Suku ke-7: 9/896
Suku ke-8: -10/1792
Suku ke-9: 11/3584


<img src="./no4-1.png" width=600/>

<img src="./no4-2.png" width=600/>

#### Nomer 5 (Deret matematika pangkat 3)

In [2]:
n = int(input("Masukkan batas maksimal suku (n): "))
jumlah = int(input("Masukkan batas maksimal total penjumlahan (jumlah): "))

suku_sekarang = 3
total_penjumlahan = 0
suku_ke = 1

print("\nDeret:")

while suku_ke <= n and total_penjumlahan + suku_sekarang <= jumlah:
    print(suku_sekarang, end=" ")
    
    total_penjumlahan = total_penjumlahan + suku_sekarang
    
    suku_sekarang = suku_sekarang * 3
    
    suku_ke = suku_ke + 1

print(f"Total penjumlahan akhir: {total_penjumlahan}")
print(f"Jumlah suku yang ditampilkan: {suku_ke - 1}")

Masukkan batas maksimal suku (n):  3
Masukkan batas maksimal total penjumlahan (jumlah):  39



Deret:
3 9 27 Total penjumlahan akhir: 39
Jumlah suku yang ditampilkan: 3


<img src="./no5.png" width=400/>

## ⚜️ Latihan

In [3]:
x = 10
y = 3
print(x%y)

1


<div style="background-color:#f5f5f5; border-left:5px solid #555; padding:12px;">
" = " Digunakan untuk assignment(penugasan) atau mendeklarasi variabel. <br>
" == " Digunakan untuk membandingkan kedua nilai.
</div>

In [7]:
n = 0

for n in range(0, 51, 5):
    print(n)

0
5
10
15
20
25
30
35
40
45
50


In [10]:
n = int(input("Masukan bilangan ganjil yang ingin ditampilkan: "))

for i in range(1, n+1):
    if(i % 2 != 0):
        print(i)
    

Masukan bilangan ganjil yang ingin ditampilkan:  8


1
3
5
7


In [12]:
i = 1
while i < 10:
    if i % 2 == 0:
        print(i)
    i+=1 #ALASAN MENGAPA TIDAK BERHENTI KARENA TIDAK TERDAPAT BARIS i+=1 PADA PERULANGAN WHILE SEHINGGA WHILE TERUS BERJALAN

2
4
6
8


### Nomer 1 (Tebak Angka)

In [18]:
komp = 6
nyawa = 3
cek_benar = False

while (nyawa > 0):
    user = int(input(f"Masukkan Tebakan ({nyawa} Kesempatan): "))

    if(user < komp):
        print("Terlalu Kecil, naikan angka mu!")
        nyawa -= 1
    elif(user > komp):
        print("Terlalu Besar, turunkan angka mu!")
        nyawa -= 1
    elif(user == komp):
        cek_benar = True
        print("\nSelamat, Jawaban Benar-!")
        break

if(cek_benar == False):
    print(f"\nGame Over, jawabannya adalah: {komp}")

Masukkan Tebakan (3 Kesempatan):  7


Terlalu Besar, turunkan angka mu!


Masukkan Tebakan (2 Kesempatan):  8


Terlalu Besar, turunkan angka mu!


Masukkan Tebakan (1 Kesempatan):  6



Selamat, Jawaban Benar-!


### Nomer 2 (Batu, Gunting, Kertas)

In [7]:
import random

list_pilihan = ["gunting", "batu", "kertas"]
komp = random.choice(list_pilihan)
batas_kesempatan = 5
user = str(input("Pilih batu/gunting/kertas:"))
skor_komp = 0
skor_user = 0

while (batas_kesempatan > 1):
    if(user == komp):
        user = str(input("Masukkan Pilihan lagi (SERI):"))
        komp = random.choice(list_pilihan)
        batas_kesempatan -= 1
    elif((user == "batu") & (komp == "kertas")):
        print("Komputer Menang = Anda Batu, Komputer Kertas")
        break
    elif((user == "gunting") & (komp == "batu")):
        print("Komputer Menang = Anda Gunting, Komputer Batu")
        break
    elif((user == "kertas") & (komp == "gunting")):
        print("Komputer Menang = Anda Kertas, Komputer Gunting")
        break
    else:
        print("Selamat Anda Menang-!", "Anda =", user, ", Komputer = ", komp)
        break

Pilih batu/gunting/kertas: batu
Masukkan Pilihan lagi (SERI): batu


Selamat Anda Menang-! Anda = batu , Komputer =  gunting


### String List

In [4]:
nilai = [50,80,65,87,52,76,90, 71,20,45]

print(nilai[0])
print(nilai[4])
print(nilai[2:7])

50
52
[65, 87, 52, 76, 90]


In [10]:
nilai = [50,80,65]
nama = ["andi", "rusdi", "amba"]

dataAll = nilai + nama

for i in range(len(dataAll)):
    print(dataAll[i])

50
80
65
andi
rusdi
amba


In [13]:
nilai = [50,84,65]
is80 = False
is84 = False

for i in range(len(nilai)):
    if(nilai[i] == 80):
        is80 = True
    elif(nilai[i] == 84):
        is84 = True

print("80 in list: ", is80)
print("84 in list: ", is84)

80 in list:  False
84 in list:  True


In [14]:
nama = ["andi", "rusdi", "amba"]
for i in range(len(nama)):
    print(nama[i])

andi
rusdi
amba


In [21]:
nama = ["andi", "rusdi", "amba"]

nama[2] = "noname"
print(nama)

['andi', 'rusdi', 'noname']


In [25]:
nilai = [51,80,65,87,52,76,90,71,20,45, 82, 65, 3, 5, 78, 32, 41,99,102, 21]
nilaigenap = []
totalgenap=0
nilaiganjil = []
totalganjil=0

for i in range(len(nilai)):
    if (nilai[i] % 2 == 0):
        nilaigenap.append(nilai[i])
        totalgenap+=nilai[i]
    else:
        nilaiganjil.append(nilai[i])
        totalganjil+=nilai[i]
print(nilaigenap, "Total: ", totalgenap)
print(nilaiganjil, "Total: ", totalganjil)

[80, 52, 76, 90, 20, 82, 78, 32, 102] Total:  612
[51, 65, 87, 71, 45, 65, 3, 5, 41, 99, 21] Total:  553


In [32]:
nilai = [51,80,65,87,52,76,90,71,20,45, 82, 65, 19, 5, 78, 32, 41,99,102, 21]
ratarata = 0

for i in range(len(nilai)):
    ratarata = sum(nilai)
    ratarata = ratarata // len(nilai)
print(ratarata)

59


In [40]:
hari = ["senin", "selasa", "rabu", "kamis", "jumat", "sabtu", "minggu"]

tanggal = int(input("Masukkan tanggal: "))
indeksawalbulan = 1

haribaru = (indeksawalbulan + tanggal) % 7
print("Hari:", hari[haribaru])

Masukkan tanggal:  22


Hari: rabu
